tools/SetupFlow/DevHome.SetupFlow/ViewModels/AddRepoViewModel.cs

```diff
@@ -581,6 +630,120 @@ public void AddRepositoryViaUri(string url, string cloneLocation)
 
         EverythingToClone.Add(cloningInformation);
+        ShouldEnablePrimaryButton = true;
+    }
+
+    /// <summary>
+    /// Tries to assign a provider to a validated uri.
+    /// </summary>
+    /// <param name="provider">The provider to test with.</param>
+    /// <param name="cloneLocation">The location the user wnats to clone the repo.</param>
+    /// <param name="uri">The uri to the repo (Should be a valid uri)</param>
+    /// <param name="loginFrame">The frame to show OAUTH login if the user needs to log in.</param>
+    /// <returns>non-null cloning information if a provider is selected for cloning.  Null for all other cases.</returns>
+    /// <remarks>If the repo is either private, or does not exist, this will ask the user to log in.</remarks>
+    private CloningInformation GetCloningInformationFromUrl(RepositoryProvider provider, string cloneLocation, Uri uri, Frame loginFrame)
+    {
+        if (provider == null)
+        {
+            // Fallback to a generic git provider.
+            // Code path lights up for a repo that has a typo.
+            var cloningInformation = new CloningInformation(new GenericRepository(uri));
+            cloningInformation.ProviderName = "git";
+            cloningInformation.CloningLocation = new DirectoryInfo(cloneLocation);
+
+            return cloningInformation;
+        }
+
+        // Repo may be public.  Try that.
+        var repo = provider.GetRepositoryFromUri(uri);
+        if (repo != null)
+        {
+            var cloningInformation = new CloningInformation(repo);
+            cloningInformation.RepositoryProvider = provider.GetProvider();
+            cloningInformation.ProviderName = provider.DisplayName;
+            cloningInformation.CloningLocation = new DirectoryInfo(cloneLocation);
+
+            return cloningInformation;
+        }
+
+        // Repo may be private, or not exist.  Try to get repo info with all logged in accounts.
+        var loggedInAccounts = provider.GetAllLoggedInAccounts();
+        if (loggedInAccounts.Any())
+        {
+            foreach (var loggedInAccount in loggedInAccounts)
+            {
+                repo = provider.GetRepositoryFromUri(uri, loggedInAccount);
+                if (repo != null)
+                {
+                    var cloningInformation = new CloningInformation(repo);
+                    cloningInformation.RepositoryProvider = provider.GetProvider();
+                    cloningInformation.ProviderName = provider.DisplayName;
+                    cloningInformation.CloningLocation = new DirectoryInfo(cloneLocation);
+                    cloningInformation.OwningAccount = loggedInAccount;
+
+                    return cloningInformation;
+                }
+            }
+
+            // In the case that no logged in accounts can access it, return null
+            // until DevHome can handle multiple accounts.
+            // Should have a better error string.

```

Comment:
Nit: use word "todo" so it's findable as a todo.

Reply:
Added.